src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,151 @@
+                return null;
+            }
+
+            // check whether there is any comments between foreach and ) tokens
+            // if they do, we don't support conversion.
+            foreach (var trivia in foreachStatement.DescendantTrivia(n => n == foreachStatement || scope.Contains(n.FullSpan)))
+            {
+                if (trivia.Span.End <= scope.Start ||
+                    scope.End <= trivia.Span.Start)
+                {
+                    continue;
+                }
+
+                if (trivia.Kind() != SyntaxKind.WhitespaceTrivia &&
+                    trivia.Kind() != SyntaxKind.EndOfLineTrivia)
+                {
+                    // we don't know what to do with these comments
+                    return null;
+                }
+            }
+
+            return foreachStatement;
+        }
+
+        protected override bool ValidLocation(ForEachInfo foreachInfo)
+        {
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                return true;
+            }
+
+            // for now, we don't support converting in embeded statement if 
+            // new local declaration for collection is required.
+            // we can support this by using Introduce local variable service
+            // but the service is not currently written in a way that can be
+            // easily reused here.
+            return foreachInfo.ForEachStatement.Parent.IsKind(SyntaxKind.Block);
+        }
+
+        protected override (SyntaxNode start, SyntaxNode end) GetForEachBody(ForEachStatementSyntax foreachStatement)
+        {
+            return (foreachStatement.Statement, foreachStatement.Statement);
+        }
+
+        protected override void ConvertToForStatement(SemanticModel model, ForEachInfo foreachInfo, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var generator = editor.Generator;
+            var foreachStatement = foreachInfo.ForEachStatement;
+
+            var foreachCollectionExpression = foreachStatement.Expression;
+            var collectionVariable = GetCollectionVariableName(model, generator, foreachInfo, foreachCollectionExpression);
+
+            // first, see whether we need to introduce new statement to capture collection
+            IntroduceCollectionStatement(model, foreachInfo, editor, foreachCollectionExpression, collectionVariable);
+
+            // create new index varialbe name
+            var indexString = CreateUniqueName(model, foreachStatement.Statement, "i");
+
+            // put variable statement in body
+            var bodyStatement = GetForLoopBody(generator, foreachInfo, collectionVariable, indexString);
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(
+                    SyntaxFactory.IdentifierName("var"),
```

Comment:
note: we need to respect user 'var' settings by default.  i don't think that can be punted till later.  users do *not* like it if they have turned off "use var" and the IDE starts introducing "var"s.  It's a very strong code style preference for many people.